File: scripts/test_production.js
#!/usr/bin/env node
/**
 * Script de pruebas E2E para verificar la web de producción
 * Prueba todas las secciones principales y verifica que no hay errores
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

// Configuración
const PRODUCTION_URL = "https://bancotest.com";
const API_URL = "https://per-api-435987927843.europe-west1.run.app";
const TIMEOUT = 30000; // 30 segundos
const SCREENSHOTS_DIR = "test_screenshots";

// Colores para output
const Colors = {
    GREEN: '\x1b[92m',
    RED: '\x1b[91m',
    YELLOW: '\x1b[93m',
    BLUE: '\x1b[94m',
    RESET: '\x1b[0m'
};

class ProductionTester {
    constructor() {
        this.results = {
            timestamp: new Date().toISOString(),
            tests: [],
            passed: 0,
            failed: 0,
            errors: []
        };
        this.browser = null;
        this.context = null;
        this.page = null;
        this.consoleErrors = [];

        // Crear directorio para screenshots
        if (!fs.existsSync(SCREENSHOTS_DIR)) {
            fs.mkdirSync(SCREENSHOTS_DIR, { recursive: true });
        }
    }

    log(message, level = "INFO") {
        let color = Colors.BLUE;
        if (level === "SUCCESS") color = Colors.GREEN;
        else if (level === "ERROR") color = Colors.RED;
        else if (level === "WARNING") color = Colors.YELLOW;

        console.log(`${color}[${level}]${Colors.RESET} ${message}`);
    }

    async startBrowser() {
        this.log("Iniciando navegador...");
        this.browser = await chromium.launch({
            headless: true
        });
        this.context = await this.browser.newContext({
            viewport: { width: 1920, height: 1080 }
        });

        this.page = await this.context.newPage();

        // Capturar errores de consola
        this.page.on('console', msg => {
            if (msg.type() === 'error' || msg.type() === 'warning') {
                const errorMsg = `${msg.type().toUpperCase()}: ${msg.text()}`;
                this.consoleErrors.push(errorMsg);
                if (msg.type() === 'error') {
                    this.log(`Console error: ${msg.text()}`, "ERROR");
                }
            }
        });

        this.page.on('pageerror', err => {
            const errorMsg = `PAGE ERROR: ${err}`;
            this.consoleErrors.push(errorMsg);
            this.log(`Page error: ${err}`, "ERROR");
        });

        this.log("Navegador iniciado", "SUCCESS");
    }

    async doLogout() {
        this.log("Haciendo logout (limpiando sesión)...");
        try {
            // Limpiar localStorage, sessionStorage y cookies
            await this.page.evaluate(() => {
                localStorage.clear();
                sessionStorage.clear();
            });
            await this.context.clearCookies();

            this.log("Sesión limpiada exitosamente", "SUCCESS");
            return true;
        } catch (error) {
            this.log(`Logout falló: ${error.message}`, "WARNING");
            return false;
        }
    }

    async doLogin() {
        const testName = "User Login";
        this.log(`Testing: ${testName}`);

        try {
            // Ir directamente a exam-system.html (requiere login)
            // Esto nos redirigirá automáticamente al login si no estamos autenticados
            await this.page.goto(`${PRODUCTION_URL}/exam-system.html`, {
                waitUntil: "networkidle",
                timeout: TIMEOUT
            });

            await this.page.waitForTimeout(3000);

            // Buscar el modal de login por el título "Iniciar Sesión"
            const loginTitle = await this.page.$('text="Iniciar Sesión"');

            if (loginTitle) {
                this.log("Página de login detectada, haciendo login...");
                await this.takeScreenshot("login_modal_detected");

                // Esperar a que los inputs sean visibles
                await this.page.waitForSelector('input[type="text"]:visible, input[type="email"]:visible', { timeout: 5000 });

                // Obtener todos los inputs visibles
                const inputs = await this.page.$$('input:visible');

                if (inputs.length >= 2) {
                    // Primer input visible = usuario/email
                    await inputs[0].fill('testuser');
                    await this.page.waitForTimeout(500);

                    // Segundo input visible = contraseña
                    await inputs[1].fill('123');
                    await this.page.waitForTimeout(500);

                    // Buscar botón "Entrar"
                    const loginBtn = await this.page.$('button:has-text("Entrar")');
                    if (loginBtn) {
                        await loginBtn.click();
                        this.log("Botón de login clickeado, esperando...");
                        await this.page.waitForTimeout(5000);

                        // Verificar si el login fue exitoso buscando el mensaje de bienvenida
                        const welcomeMessage = await this.page.$('text=/¡Bienvenido|Bienvenido|Nuevo Examen/i');

                        if (welcomeMessage) {
                            this.log("Login exitoso - Bienvenida detectada", "SUCCESS");
                            await this.takeScreenshot("logged_in");
                            return true;
                        }

                        // Si no hay mensaje de bienvenida, verificar si seguimos en el modal de login
                        const stillLoginModal = await this.page.$('text="Iniciar Sesión"');

                        if (stillLoginModal) {
                            this.log("Usuario no existe o credenciales incorrectas, intentando registrar...", "WARNING");

                            // Buscar link de registro
                            const registerLink = await this.page.$('a:has-text("Regístrate")');
                            if (registerLink) {
                                await registerLink.click();
                                await this.page.waitForTimeout(2000);

                                // Esperar modal de registro
                                await this.page.waitForSelector('input:visible', { timeout: 5000 });
                                const registerInputs = await this.page.$$('input:visible');

                                if (registerInputs.length >= 3) {
                                    // Nombre de usuario
                                    await registerInputs[0].fill('testuser');
                                    await this.page.waitForTimeout(500);

                                    // Email
                                    await registerInputs[1].fill('testuser@test.com');
                                    await this.page.waitForTimeout(500);

                                    // Contraseña
                                    await registerInputs[2].fill('123');
                                    await this.page.waitForTimeout(500);

                                    // Buscar botón de registro
                                    const registerBtn = await this.page.$('button:has-text("Registrar"), button:has-text("Crear")');
                                    if (registerBtn) {
                                        await registerBtn.click();
                                        this.log("Usuario registrado, esperando...");
                                        await this.page.waitForTimeout(5000);
                                    }
                                }
                            }
                        }

                        this.log("Login completado", "SUCCESS");
                        await this.takeScreenshot("logged_in");
                        return true;
                    } else {
                        this.log("No se encontró botón 'Entrar'", "ERROR");
                        await this.takeScreenshot("no_login_button");
                        return false;
                    }
                } else {
                    this.log("No se encontraron suficientes inputs visibles", "ERROR");
                    await this.takeScreenshot("no_inputs");
                    return false;
                }
            } else {
                this.log("Ya está autenticado (no aparece modal de login)", "SUCCESS");
                await this.takeScreenshot("already_logged_in");
                return true;
            }

        } catch (error) {
            this.log(`Login falló: ${error.message}`, "ERROR");
            await this.takeScreenshot("login_error");
            return false;
        }
    }

    async takeScreenshot(name) {
        const filename = `${SCREENSHOTS_DIR}/${name}_${Date.now()}.png`;
        await this.page.screenshot({ path: filename });
        return filename;
    }

    async testApiHealth() {
        const testName = "API Health Check";
        this.log(`Testing: ${testName}`);

        try {
            const response = await this.page.request.get(`${API_URL}/health`);
            if (response.status() !== 200) {
                throw new Error(`API health check failed with status ${response.status()}`);
            }

            const data = await response.json();
            if (data.status !== "healthy") {
                throw new Error("API status is not healthy");
            }
            if (data.database !== "connected") {
                throw new Error("Database is not connected");
            }

            this.recordSuccess(testName);
            return true;
        } catch (error) {
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testHomepage() {
        const testName = "Homepage";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];
            await this.page.goto(PRODUCTION_URL, {
                waitUntil: "networkidle",
                timeout: TIMEOUT
            });

            // Verificar título
            const title = await this.page.title();
            if (!title.includes("PER")) {
                throw new Error(`Unexpected title: ${title}`);
            }

            // Verificar que no hay errores de consola críticos
            const criticalErrors = this.consoleErrors.filter(e => e.includes("ERROR:"));
            if (criticalErrors.length > 0) {
                throw new Error(`Console errors found: ${criticalErrors.join(", ")}`);
            }

            await this.takeScreenshot("homepage");
            this.recordSuccess(testName);
            return true;
        } catch (error) {
            await this.takeScreenshot("homepage_error");
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testExamSystem() {
        const testName = "Exam System (Nuevo Examen)";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];

            // Si estamos en el dashboard, hacer click en "Nuevo Examen"
            const newExamBtn = await this.page.$('text="Nuevo Examen"');
            if (newExamBtn) {
                await newExamBtn.click();
                await this.page.waitForTimeout(3000);
            } else {
                throw new Error("No se encontró botón 'Nuevo Examen' en dashboard");
            }

            // Verificar que se cargó un examen (buscar elementos que indican que hay una pregunta)
            const hasQuestion = await this.page.$('text=/Pregunta \\d+ de|pregunta/i');
            const hasAnswerOptions = await this.page.$$('input[type="radio"]');

            if (!hasQuestion) {
                throw new Error("No se detectó texto de pregunta");
            }

            if (hasAnswerOptions.length < 2) {
                throw new Error("No se detectaron opciones de respuesta");
            }

            // Verificar que no hay errores críticos
            const criticalErrors = this.consoleErrors.filter(e =>
                e.includes("ERROR:") && !e.includes("404")
            );
            if (criticalErrors.length > 0) {
                throw new Error(`Console errors found: ${criticalErrors.join(", ")}`);
            }

            await this.takeScreenshot("exam_system_success");
            this.recordSuccess(testName);
            return true;
        } catch (error) {
            await this.takeScreenshot("exam_system_error");
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testQuestionStatistics() {
        const testName = "Estadísticas de Preguntas";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];

            // Hacer click en "Estadísticas de Preguntas"
            const statsBtn = await this.page.$('text="Estadísticas de Preguntas"');
            if (statsBtn) {
                await statsBtn.click();
                await this.page.waitForTimeout(3000);
            } else {
                throw new Error("No se encontró botón 'Estadísticas de Preguntas'");
            }

            // Verificar que se cargó la página de estadísticas
            // Puede tener diferentes contenidos dependiendo de si hay datos
            const hasContent = await this.page.$('body');

            if (!hasContent) {
                throw new Error("La página no cargó contenido");
            }

            // Verificar que no hay errores 500 o críticos
            const criticalErrors = this.consoleErrors.filter(e =>
                e.includes("ERROR:") && (e.includes("500") || e.includes("Failed"))
            );
            if (criticalErrors.length > 0) {
                throw new Error(`Errors found: ${criticalErrors.join(", ")}`);
            }

            await this.takeScreenshot("question_statistics_success");
            this.recordSuccess(testName);
            return true;
        } catch (error) {
            await this.takeScreenshot("question_statistics_error");
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testStatisticsDashboard() {
        const testName = "Estadísticas (Dashboard)";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];

            // Hacer click en "Estadísticas"
            const statsBtn = await this.page.$('text=/^Estadísticas$/');
            if (statsBtn) {
                await statsBtn.click();
                await this.page.waitForTimeout(3000);
            } else {
                // Alternativa: navegar directamente
                await this.page.goto(`${PRODUCTION_URL}/statistics-dashboard.html`, {
                    waitUntil: "networkidle",
                    timeout: TIMEOUT
                });
            }

            // Verificar que no hay errores 500 en user-stats
            const criticalErrors = this.consoleErrors.filter(e =>
                e.includes("500") || e.includes("Failed to load user statistics")
            );

            if (criticalErrors.length > 0) {
                throw new Error(`Statistics loading errors: ${criticalErrors.join(", ")}`);
            }

            await this.takeScreenshot("statistics_dashboard_success");
            this.recordSuccess(testName);
            return true;
        } catch (error) {
            await this.takeScreenshot("statistics_dashboard_error");
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testQuestionBrowser() {
        const testName = "Banco de Preguntas";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];

            // Hacer click en "Banco de Preguntas"
            const bankBtn = await this.page.$('text="Banco de Preguntas"');
            if (bankBtn) {
                await bankBtn.click();
                await this.page.waitForTimeout(3000);
            } else {
                throw new Error("No se encontró botón 'Banco de Preguntas'");
            }

            // Verificar que la página cargó (puede ser index.html o visor)
            // Simplemente verificamos que no hay errores críticos
            const criticalErrors = this.consoleErrors.filter(e =>
                e.includes("ERROR:") && (e.includes("500") || e.includes("Failed"))
            );

            if (criticalErrors.length > 0) {
                throw new Error(`Errors found: ${criticalErrors.join(", ")}`);
            }

            await this.takeScreenshot("question_browser_success");
            this.recordSuccess(testName);
            return true;
        } catch (error) {
            await this.takeScreenshot("question_browser_error");
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testAdminPanel() {
        const testName = "Panel de Administración";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];

            // Hacer click en "Panel de Administración"
            const adminBtn = await this.page.$('text="Panel de Administración"');
            if (adminBtn) {
                await adminBtn.click();
                await this.page.waitForTimeout(3000);
            } else {
                throw new Error("No se encontró botón 'Panel de Administración'");
            }

            // Verificar que no hay errores críticos
            const criticalErrors = this.consoleErrors.filter(e =>
                e.includes("ERROR:") && (e.includes("500") || e.includes("Failed"))
            );

            if (criticalErrors.length > 0) {
                throw new Error(`Errors found: ${criticalErrors.join(", ")}`);
            }

            await this.takeScreenshot("admin_panel_success");
            this.recordSuccess(testName);
            return true;
        } catch (error) {
            await this.takeScreenshot("admin_panel_error");
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    async testCorsFunctionality() {
        const testName = "CORS Functionality";
        this.log(`Testing: ${testName}`);

        try {
            this.consoleErrors = [];
            await this.page.goto(PRODUCTION_URL, {
                waitUntil: "networkidle",
                timeout: TIMEOUT
            });

            // Hacer una petición a la API desde el frontend
            const response = await this.page.evaluate(async (apiUrl) => {
                try {
                    const res = await fetch(`${apiUrl}/health`);
                    return await res.json();
                } catch (err) {
                    return { error: err.message };
                }
            }, API_URL);

            if (response.error) {
                throw new Error(`CORS error: ${response.error}`);
            }
            if (response.status !== "healthy") {
                throw new Error("API not healthy");
            }

            // Verificar que no hay errores CORS en consola
            const corsErrors = this.consoleErrors.filter(e => e.toUpperCase().includes("CORS"));
            if (corsErrors.length > 0) {
                throw new Error(`CORS errors found: ${corsErrors.join(", ")}`);
            }

            this.recordSuccess(testName);
            return true;
        } catch (error) {
            this.recordFailure(testName, error.message);
            return false;
        }
    }

    recordSuccess(testName) {
        this.results.tests.push({
            name: testName,
            status: "PASSED",
            timestamp: new Date().toISOString()
        });
        this.results.passed++;
        this.log(`✓ ${testName} PASSED`, "SUCCESS");
    }

    recordFailure(testName, error) {
        this.results.tests.push({
            name: testName,
            status: "FAILED",
            error: error,
            timestamp: new Date().toISOString()
        });
        this.results.failed++;
        this.results.errors.push(`${testName}: ${error}`);
        this.log(`✗ ${testName} FAILED: ${error}`, "ERROR");
    }

    async cleanup() {
        if (this.context) await this.context.close();
        if (this.browser) await this.browser.close();
    }

    saveResults() {
        const filename = `test_results_${Date.now()}.json`;
        fs.writeFileSync(filename, JSON.stringify(this.results, null, 2));
        this.log(`Results saved to ${filename}`, "SUCCESS");
    }

    printSummary() {
        console.log("\n" + "=".repeat(60));
        console.log("TEST SUMMARY");
        console.log("=".repeat(60));
        console.log(`Total tests: ${this.results.passed + this.results.failed}`);
        console.log(`${Colors.GREEN}Passed: ${this.results.passed}${Colors.RESET}`);
        console.log(`${Colors.RED}Failed: ${this.results.failed}${Colors.RESET}`);
        console.log("=".repeat(60));

        if (this.results.errors.length > 0) {
            console.log(`\n${Colors.RED}ERRORS:${Colors.RESET}`);
            this.results.errors.forEach(error => {
                console.log(`  - ${error}`);
            });
        }

        console.log(`\nScreenshots saved in: ${SCREENSHOTS_DIR}/`);
    }

    async runAllTests() {
        try {
            await this.startBrowser();

            // Tests sin autenticación
            const publicTests = [
                () => this.testApiHealth(),
                () => this.testHomepage(),
                () => this.testCorsFunctionality(),
            ];

            // Ejecutar tests públicos
            for (const testFunc of publicTests) {
                try {
                    await testFunc();
                } catch (error) {
                    this.log(`Unexpected error in test: ${error.message}`, "ERROR");
                }
                await new Promise(resolve => setTimeout(resolve, 1000));
            }

            // Tests que requieren autenticación (verifican las 5 secciones principales del menú)
            const authTests = [
                { name: "Nuevo Examen", func: () => this.testExamSystem() },
                { name: "Estadísticas de Preguntas", func: () => this.testQuestionStatistics() },
                { name: "Estadísticas", func: () => this.testStatisticsDashboard() },
                { name: "Banco de Preguntas", func: () => this.testQuestionBrowser() },
                { name: "Panel de Administración", func: () => this.testAdminPanel() },
            ];

            // Ejecutar cada test autenticado con login/logout
            for (const test of authTests) {
                try {
                    // Login antes del test
                    await this.doLogin();

                    // Ejecutar el test
                    await test.func();

                    // Logout después del test
                    await this.doLogout();

                } catch (error) {
                    this.log(`Unexpected error in ${test.name}: ${error.message}`, "ERROR");
                    // Intentar logout aunque falle el test
                    try {
                        await this.doLogout();
                    } catch (e) {
                        this.log(`Logout falló después de error en test`, "WARNING");
                    }
                }
                await new Promise(resolve => setTimeout(resolve, 1000));
            }

        } finally {
            await this.cleanup();
            this.saveResults();
            this.printSummary();
        }

        // Retornar código de salida
        return this.results.failed === 0 ? 0 : 1;
    }
}

async function main() {
    console.log(`\n${Colors.BLUE}${"=".repeat(60)}${Colors.RESET}`);
    console.log(`${Colors.BLUE}PER Production Testing Suite${Colors.RESET}`);
    console.log(`${Colors.BLUE}${"=".repeat(60)}${Colors.RESET}\n`);
    console.log(`Testing URL: ${PRODUCTION_URL}`);
    console.log(`API URL: ${API_URL}\n`);

    const tester = new ProductionTester();
    const exitCode = await tester.runAllTests();

    process.exit(exitCode);
}

main();

File: scripts/test_production.py
#!/usr/bin/env python3
"""
Script de pruebas E2E para verificar la web de producción
Prueba todas las secciones principales y verifica que no hay errores
"""

import os
import sys
import time
import json
from datetime import datetime
from playwright.sync_api import sync_playwright, expect

# Configuración
PRODUCTION_URL = "https://bancotest.com"
API_URL = "https://per-api-435987927843.europe-west1.run.app"
TIMEOUT = 30000  # 30 segundos
SCREENSHOTS_DIR = "test_screenshots"

# Colores para output
class Colors:
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    RESET = '\033[0m'

class ProductionTester:
    def __init__(self):
        self.results = {
            "timestamp": datetime.now().isoformat(),
            "tests": [],
            "passed": 0,
            "failed": 0,
            "errors": []
        }
        self.browser = None
        self.context = None
        self.page = None

        # Crear directorio para screenshots
        os.makedirs(SCREENSHOTS_DIR, exist_ok=True)

    def log(self, message, level="INFO"):
        color = Colors.BLUE
        if level == "SUCCESS":
            color = Colors.GREEN
        elif level == "ERROR":
            color = Colors.RED
        elif level == "WARNING":
            color = Colors.YELLOW

        print(f"{color}[{level}]{Colors.RESET} {message}")

    def start_browser(self):
        """Iniciar navegador"""
        self.log("Iniciando navegador...")
        playwright = sync_playwright().start()
        self.browser = playwright.chromium.launch(headless=True)
        self.context = self.browser.new_context(
            viewport={'width': 1920, 'height': 1080}
        )

        # Capturar errores de consola
        self.console_errors = []
        self.page = self.context.new_page()
        self.page.on("console", lambda msg: self._handle_console(msg))
        self.page.on("pageerror", lambda err: self._handle_page_error(err))

        self.log("Navegador iniciado", "SUCCESS")

    def _handle_console(self, msg):
        """Capturar mensajes de consola"""
        if msg.type in ["error", "warning"]:
            error_msg = f"{msg.type.upper()}: {msg.text}"
            self.console_errors.append(error_msg)
            if msg.type == "error":
                self.log(f"Console error: {msg.text}", "ERROR")

    def _handle_page_error(self, err):
        """Capturar errores de página"""
        error_msg = f"PAGE ERROR: {err}"
        self.console_errors.append(error_msg)
        self.log(f"Page error: {err}", "ERROR")

    def take_screenshot(self, name):
        """Tomar screenshot"""
        filename = f"{SCREENSHOTS_DIR}/{name}_{int(time.time())}.png"
        self.page.screenshot(path=filename)
        return filename

    def test_api_health(self):
        """Verificar que la API está funcionando"""
        test_name = "API Health Check"
        self.log(f"Testing: {test_name}")

        try:
            response = self.page.request.get(f"{API_URL}/health")
            assert response.status == 200, f"API health check failed with status {response.status}"

            data = response.json()
            assert data.get("status") == "healthy", "API status is not healthy"
            assert data.get("database") == "connected", "Database is not connected"

            self._record_success(test_name)
            return True
        except Exception as e:
            self._record_failure(test_name, str(e))
            return False

    def test_homepage(self):
        """Verificar página principal"""
        test_name = "Homepage"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(PRODUCTION_URL, wait_until="networkidle", timeout=TIMEOUT)

            # Verificar título
            expect(self.page).to_have_title("PER Examen - Sistema de Exámenes")

            # Verificar que no hay errores de consola críticos
            critical_errors = [e for e in self.console_errors if "ERROR:" in e]
            if critical_errors:
                raise Exception(f"Console errors found: {critical_errors}")

            self.take_screenshot("homepage")
            self._record_success(test_name)
            return True
        except Exception as e:
            self.take_screenshot(f"homepage_error")
            self._record_failure(test_name, str(e))
            return False

    def test_login(self):
        """Verificar login de usuario"""
        test_name = "User Login"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(PRODUCTION_URL, wait_until="networkidle", timeout=TIMEOUT)

            # Click en botón de login
            self.page.click("button:has-text('Acceder')")
            self.page.wait_for_timeout(1000)

            # Rellenar formulario de login (usuario de prueba)
            self.page.fill("input[type='email']", "test@example.com")
            self.page.fill("input[type='password']", "test123")

            # Hacer login
            self.page.click("button:has-text('Iniciar')")
            self.page.wait_for_timeout(2000)

            # Si no existe el usuario, registrarlo
            if "Registro" in self.page.content() or "Register" in self.page.content():
                self.log("Usuario no existe, registrando...", "WARNING")
                # Hacer registro primero
                self.page.goto(PRODUCTION_URL)
                self.page.click("button:has-text('Registrarse')")
                self.page.wait_for_timeout(1000)

                self.page.fill("input[type='text']", "Test User")
                self.page.fill("input[type='email']", "test@example.com")
                self.page.fill("input[type='password']", "test123")
                self.page.click("button:has-text('Crear')")
                self.page.wait_for_timeout(2000)

            # Verificar que estamos logueados (buscar elementos típicos de usuario logueado)
            self.page.wait_for_timeout(2000)

            self.take_screenshot("logged_in")
            self._record_success(test_name)
            return True
        except Exception as e:
            self.take_screenshot(f"login_error")
            self._record_failure(test_name, str(e))
            return False

    def test_exam_system(self):
        """Verificar sistema de exámenes"""
        test_name = "Exam System"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(f"{PRODUCTION_URL}/exam-system.html", wait_until="networkidle", timeout=TIMEOUT)

            # Esperar a que cargue la configuración
            self.page.wait_for_selector("#questionCount", timeout=10000)

            # Verificar elementos principales
            assert self.page.is_visible("#questionCount"), "Question count selector not visible"
            assert self.page.is_visible("#startExamBtn"), "Start exam button not visible"

            # Verificar que no hay errores críticos
            critical_errors = [e for e in self.console_errors if "ERROR:" in e and "404" not in e]
            if critical_errors:
                raise Exception(f"Console errors found: {critical_errors}")

            self.take_screenshot("exam_system")
            self._record_success(test_name)
            return True
        except Exception as e:
            self.take_screenshot(f"exam_system_error")
            self._record_failure(test_name, str(e))
            return False

    def test_create_exam(self):
        """Verificar creación de examen"""
        test_name = "Create Exam"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(f"{PRODUCTION_URL}/exam-system.html", wait_until="networkidle", timeout=TIMEOUT)

            # Esperar a que cargue
            self.page.wait_for_selector("#startExamBtn", timeout=10000)

            # Seleccionar 10 preguntas
            self.page.select_option("#questionCount", "10")
            self.page.wait_for_timeout(500)

            # Iniciar examen
            self.page.click("#startExamBtn")

            # Esperar a que aparezca la primera pregunta
            self.page.wait_for_selector("#questionText", timeout=15000)

            # Verificar que hay opciones de respuesta
            assert self.page.is_visible(".answer-option"), "Answer options not visible"

            self.take_screenshot("exam_created")
            self._record_success(test_name)
            return True
        except Exception as e:
            self.take_screenshot(f"create_exam_error")
            self._record_failure(test_name, str(e))
            return False

    def test_statistics_dashboard(self):
        """Verificar dashboard de estadísticas"""
        test_name = "Statistics Dashboard"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(f"{PRODUCTION_URL}/statistics-dashboard.html", wait_until="networkidle", timeout=TIMEOUT)

            # Esperar a que cargue el dashboard
            self.page.wait_for_timeout(3000)

            # Verificar que el dashboard se muestra
            # (puede estar vacío si no hay datos, pero no debería dar errores)

            # Verificar que no hay errores 500 o 404 en user-stats
            critical_errors = [e for e in self.console_errors
                             if ("500" in e or "Failed to load user statistics" in e)]

            if critical_errors:
                raise Exception(f"Statistics loading errors: {critical_errors}")

            self.take_screenshot("statistics_dashboard")
            self._record_success(test_name)
            return True
        except Exception as e:
            self.take_screenshot(f"statistics_error")
            self._record_failure(test_name, str(e))
            return False

    def test_question_browser(self):
        """Verificar navegador de preguntas"""
        test_name = "Question Browser"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(f"{PRODUCTION_URL}/index.html", wait_until="networkidle", timeout=TIMEOUT)

            # Esperar a que carguen las preguntas
            self.page.wait_for_selector(".question-item", timeout=10000)

            # Verificar que hay preguntas
            questions = self.page.query_selector_all(".question-item")
            assert len(questions) > 0, "No questions loaded"

            # Click en una pregunta
            questions[0].click()
            self.page.wait_for_timeout(1000)

            # Verificar que se muestra el detalle
            assert self.page.is_visible(".question-detail"), "Question detail not visible"

            self.take_screenshot("question_browser")
            self._record_success(test_name)
            return True
        except Exception as e:
            self.take_screenshot(f"question_browser_error")
            self._record_failure(test_name, str(e))
            return False

    def test_cors_functionality(self):
        """Verificar que CORS funciona correctamente"""
        test_name = "CORS Functionality"
        self.log(f"Testing: {test_name}")

        try:
            self.console_errors = []
            self.page.goto(PRODUCTION_URL, wait_until="networkidle", timeout=TIMEOUT)

            # Hacer una petición a la API desde el frontend
            response = self.page.evaluate(f"""
                fetch('{API_URL}/health')
                    .then(r => r.json())
                    .then(data => data)
                    .catch(err => ({{ error: err.message }}))
            """)

            assert "error" not in response, f"CORS error: {response.get('error')}"
            assert response.get("status") == "healthy", "API not healthy"

            # Verificar que no hay errores CORS en consola
            cors_errors = [e for e in self.console_errors if "CORS" in e.upper()]
            if cors_errors:
                raise Exception(f"CORS errors found: {cors_errors}")

            self._record_success(test_name)
            return True
        except Exception as e:
            self._record_failure(test_name, str(e))
            return False

    def _record_success(self, test_name):
        """Registrar test exitoso"""
        self.results["tests"].append({
            "name": test_name,
            "status": "PASSED",
            "timestamp": datetime.now().isoformat()
        })
        self.results["passed"] += 1
        self.log(f"✓ {test_name} PASSED", "SUCCESS")

    def _record_failure(self, test_name, error):
        """Registrar test fallido"""
        self.results["tests"].append({
            "name": test_name,
            "status": "FAILED",
            "error": error,
            "timestamp": datetime.now().isoformat()
        })
        self.results["failed"] += 1
        self.results["errors"].append(f"{test_name}: {error}")
        self.log(f"✗ {test_name} FAILED: {error}", "ERROR")

    def cleanup(self):
        """Limpiar recursos"""
        if self.context:
            self.context.close()
        if self.browser:
            self.browser.close()

    def save_results(self):
        """Guardar resultados en JSON"""
        filename = f"test_results_{int(time.time())}.json"
        with open(filename, 'w') as f:
            json.dump(self.results, f, indent=2)
        self.log(f"Results saved to {filename}", "SUCCESS")

    def print_summary(self):
        """Imprimir resumen de resultados"""
        print("\n" + "="*60)
        print("TEST SUMMARY")
        print("="*60)
        print(f"Total tests: {self.results['passed'] + self.results['failed']}")
        print(f"{Colors.GREEN}Passed: {self.results['passed']}{Colors.RESET}")
        print(f"{Colors.RED}Failed: {self.results['failed']}{Colors.RESET}")
        print("="*60)

        if self.results['errors']:
            print(f"\n{Colors.RED}ERRORS:{Colors.RESET}")
            for error in self.results['errors']:
                print(f"  - {error}")

        print(f"\nScreenshots saved in: {SCREENSHOTS_DIR}/")

    def run_all_tests(self):
        """Ejecutar todos los tests"""
        try:
            self.start_browser()

            # Lista de tests a ejecutar
            tests = [
                self.test_api_health,
                self.test_homepage,
                self.test_cors_functionality,
                self.test_exam_system,
                self.test_create_exam,
                self.test_statistics_dashboard,
                self.test_question_browser,
            ]

            # Ejecutar cada test
            for test_func in tests:
                try:
                    test_func()
                except Exception as e:
                    self.log(f"Unexpected error in {test_func.__name__}: {e}", "ERROR")

                # Pequeña pausa entre tests
                time.sleep(1)

        finally:
            self.cleanup()
            self.save_results()
            self.print_summary()

        # Retornar código de salida
        return 0 if self.results['failed'] == 0 else 1

def main():
    """Main function"""
    print(f"\n{Colors.BLUE}{'='*60}{Colors.RESET}")
    print(f"{Colors.BLUE}PER Production Testing Suite{Colors.RESET}")
    print(f"{Colors.BLUE}{'='*60}{Colors.RESET}\n")
    print(f"Testing URL: {PRODUCTION_URL}")
    print(f"API URL: {API_URL}\n")

    tester = ProductionTester()
    exit_code = tester.run_all_tests()

    sys.exit(exit_code)

if __name__ == "__main__":
    main()
